engine/sources/httpdns/source.go

```diff
@@ -27,128 +25,91 @@ import (
 	"github.com/moby/buildkit/solver/pb"
 	"github.com/moby/buildkit/source"
 	srchttp "github.com/moby/buildkit/source/http"
-	srctypes "github.com/moby/buildkit/source/types"
 	"github.com/moby/buildkit/util/tracing"
-	"github.com/moby/locker"
 	digest "github.com/opencontainers/go-digest"
 	"github.com/pkg/errors"
 )
 
 type Opt struct {
-	CacheAccessor cache.Accessor
+	srchttp.Opt
 	BaseDNSConfig *oci.DNSConfig
-	Transport     http.RoundTripper
 }
 
 type httpSource struct {
+	src source.Source
+
 	cache     cache.Accessor
-	locker    *locker.Locker
 	transport http.RoundTripper
 	dns       *oci.DNSConfig
 }
 
 func NewSource(opt Opt) (source.Source, error) {
+	src, err := srchttp.NewSource(opt.Opt)
+	if err != nil {
+		return nil, err
+	}
+
 	transport := opt.Transport
 	if transport == nil {
 		transport = tracing.DefaultTransport
 	}
 	hs := &httpSource{
+		src:       src,
 		cache:     opt.CacheAccessor,
-		locker:    locker.New(),
 		transport: transport,
 		dns:       opt.BaseDNSConfig,
 	}
 	return hs, nil
 }
 
 func (hs *httpSource) Schemes() []string {
-	return []string{srctypes.HTTPSScheme}
+	return hs.src.Schemes()
 }
 
 func (hs *httpSource) Identifier(scheme, ref string, attrs map[string]string, platform *pb.Platform) (source.Identifier, error) {
-	id, err := srchttp.NewHTTPIdentifier(ref, scheme == "https")
+	srcid, err := hs.src.Identifier(scheme, ref, attrs, platform)
 	if err != nil {
 		return nil, err
 	}
+	id := &HTTPIdentifier{
+		HTTPIdentifier: *(srcid.(*srchttp.HTTPIdentifier)),
+	}
 
+	//nolint:gocritic
 	for k, v := range attrs {
 		switch k {
-		case pb.AttrHTTPChecksum:
-			dgst, err := digest.Parse(v)
-			if err != nil {
-				return nil, err
-			}
-			id.Checksum = dgst
-		case pb.AttrHTTPFilename:
-			id.Filename = v
-		case pb.AttrHTTPPerm:
-			i, err := strconv.ParseInt(v, 0, 64)
-			if err != nil {
-				return nil, err
-			}
-			id.Perm = int(i)
-		case pb.AttrHTTPUID:
-			i, err := strconv.ParseInt(v, 0, 64)
-			if err != nil {
-				return nil, err
-			}
-			id.UID = int(i)
-		case pb.AttrHTTPGID:
-			i, err := strconv.ParseInt(v, 0, 64)
-			if err != nil {
-				return nil, err
-			}
-			id.GID = int(i)
+		case "dagger.git.clientids":
```

Comment:
nit: might as well const this. Also looks like we don't need the loop anymore.

edit: oh also this probably shouldn't be `git` :D